### PR TITLE
nested navigation menu items

### PR DESCRIPTION
**Pull request by sadaf895**
#2249

### CODE REVIEW

File: src/app/core/config/dynamic-routing/route-permissions.service.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export class RoutePermissionsService {
   async filterPermittedRoutes(items: MenuItem[]): Promise<MenuItem[]> {
     const accessibleRoutes: MenuItem[] = [];
     for (const item of items) {
-      if (await this.isAccessibleRouteForUser(item.link)) {
+      if (await this.isAccessibleRouteForUser(item.link as string)) {
```
Comment (sleidig) [inline src/app/core/config/dynamic-routing/route-permissions.service.ts:24]: This is slightly tricky: We check and remove menu items here if the current user doesn't have permissions to access the link.

For submenu sections we'll have to check the subMenu items instead of the item itself (which doesn't have a link, right?) and
- [ ] add the parent menu item only if at least one of the submenu items is accessible
- [ ] check each submenu item and only show those that are accessible to the current user

Comment (sleidig) [reply src/app/core/config/dynamic-routing/route-permissions.service.ts:24]: maybe instead of doing this here, we can move it into the individual `MenuItemComponent`.

**Review by sleidig — COMMENTED** [review]
@sadaf895 , let us look at the permission aspects together in a joint session

File: src/app/core/config/dynamic-routing/route-permissions.service.ts [hunk 2]
```diff
@@ -20,15 +20,27 @@ export class RoutePermissionsService {
...
-      if (await this.isAccessibleRouteForUser(item.link)) {
+      if (item.link && (await this.isAccessibleRouteForUser(item.link))) {
         accessibleRoutes.push(item);
+      } else if (item.subMenu) {
+        const accessibleSubItems: MenuItem[] = await this.filterPermittedRoutes(
+          item.subMenu,
+        );
+
+        if (accessibleSubItems.length > 0) {
+          const filteredParentItem: MenuItem = Object.assign({}, item);
```
Comment (sleidig) [inline src/app/core/config/dynamic-routing/route-permissions.service.ts:33]:
```suggestion
          // only adding the item if there is at least one accessible subMenu item
          const filteredParentItem: MenuItem = Object.assign({}, item);
```

File: src/app/core/ui/navigation/menu-item.component.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    RouterLink,
+    Angulartics2Module,
+    NgForOf,
+    MatMenuModule,
+  ],
+  standalone: true,
+})
+export class MenuItemComponent {
+  @Input() item: MenuItem; // Receives each individual menu item as input
+  activeLink: string;
```
Comment (sleidig) [inline src/app/core/ui/navigation/menu-item.component.ts:30]: what is "activeLink" used for?

File: src/app/core/ui/navigation/navigation/navigation.component.ts [hunk 4]
```diff
@@ -26,9 +26,11 @@ import { NgForOf } from "@angular/common";
 import { Angulartics2Module } from "angulartics2";
 import { FaDynamicIconComponent } from "../../../common-components/fa-dynamic-icon/fa-dynamic-icon.component";
 import { RoutePermissionsService } from "../../../config/dynamic-routing/route-permissions.service";
+import { MatMenuModule } from "@angular/material/menu";
+import { CommonModule } from "@angular/common";
+import { MenuItemComponent } from "../menu-item.component";
 
-/**
- * Main app menu listing.
+/* Main app menu listing.
```
Comment (sleidig) [inline src/app/core/ui/navigation/navigation/navigation.component.ts:33]:
```suggestion
/**
 * Main app menu listing.
```

File: src/app/core/ui/navigation/navigation/navigation.component.ts [hunk 5]
```diff
@@ -82,16 +88,16 @@ export class NavigationComponent {
...
-      newUrl.startsWith(item.link),
+    const items: MenuItem[] = this.menuItems.filter(
+      (item) => typeof item.link === "string" && newUrl.startsWith(item.link),
     );
     switch (items.length) {
       case 0:
         return "";
       case 1:
-        const link = items[0].link;
+        const target = items[0].link as string;
```
Comment (sleidig) [inline src/app/core/ui/navigation/navigation/navigation.component.ts:98]: @sadaf895 , I think we don't need to rename this anymore. Do you want to undo these changes and explicit types ("as string") to keep the changes to a minimum?

File: src/app/core/ui/navigation/menu-item.component.ts (hunk 3 above)
Comment (sadaf895) [reply src/app/core/ui/navigation/menu-item.component.ts:30]: To keep track of the currently active route or link within the menu.

Comment (sadaf895) [reply src/app/core/ui/navigation/menu-item.component.ts]: Yes, right  I should pass the activeLink value as an @Input to MenuItemComponent.

**Comment by sadaf895** [comment]
It is working for me now, @sleidig Let me know that whether it is working well for you also.

File: src/app/core/config/dynamic-routing/route-permissions.service.spec.ts [hunk 6]
```diff
@@ -20,4 +38,86 @@ describe("RoutePermissionsService", () => {
...
+
+Nested:
+  item 1
+    1.1
+    1.2   x
+
+  item 2  x
+    2.1   x
+
+ */
```
Comment (sleidig) [inline src/app/core/config/dynamic-routing/route-permissions.service.spec.ts:123]:
```suggestion
 */

```

File: src/app/core/ui/navigation/menu-item.component.ts (hunk 3 above)
Comment (sleidig) [reply src/app/core/ui/navigation/menu-item.component.ts]: > It is working for me now, Let me know that whether it is working well for you also.

Nope, not working for me @sadaf895 . Did you run the app and see the menu item highlighted correctly? We need to get back to a design where the currently visited menu is white background (see screenshots)

In the code I don't see you passing in any value to the `activeMenuItem` input also.
